src/query/src/within_filter.rs

```diff
@@ -0,0 +1,359 @@
+        assert!(catalog_list
+            .register_table_sync(RegisterTableRequest {
+                catalog: DEFAULT_CATALOG_NAME.to_string(),
+                schema: DEFAULT_SCHEMA_NAME.to_string(),
+                table_name,
+                table_id: 1024,
+                table,
+            })
+            .is_ok());
+        QueryEngineFactory::new(catalog_list, None, None, None, None, false).query_engine()
+    }
+
+    async fn do_query(sql: &str) -> Result<LogicalPlan> {
+        let stmt = QueryLanguageParser::parse_sql(sql, &QueryContext::arc()).unwrap();
+        let engine = create_test_engine().await;
+        engine.planner().plan(&stmt, QueryContext::arc()).await
+    }
+
+    #[tokio::test]
+    async fn test_within_filter() {
+        // TODO: test within filter with time zone
+
+        // 2015-01-01T00:00:00 <= timestamp < 2016-01-01T00:00:00
+        let sql = "SELECT * FROM test WHERE ts WITHIN '2015'";
+        let plan = do_query(sql).await.unwrap();
+        let expected = "Projection: *\
+        \n  Filter: test.ts >= TimestampSecond(1420070400, None) AND test.ts < TimestampSecond(1451606400, None)\
+        \n    TableScan: test";
+        assert_eq!(expected, plan.to_string());
+
+        // 2025-03-01T00:00:00 <= timestamp < 2025-04-01T00:00:00
+        let sql = "SELECT * FROM test WHERE ts WITHIN '2025-3'";
+        let plan = do_query(sql).await.unwrap();
+        let expected = "Projection: *\
+        \n  Filter: test.ts >= TimestampSecond(1740787200, None) AND test.ts < TimestampSecond(1743465600, None)\
+        \n    TableScan: test";
+        assert_eq!(expected, plan.to_string());
+
+        // 2025-12-1T00:00:00 <= timestamp < 2026-01-01T00:00:00
+        let sql = "SELECT * FROM test WHERE ts WITHIN '2025-12'";
+        let plan = do_query(sql).await.unwrap();
+        let expected = "Projection: *\
+        \n  Filter: test.ts >= TimestampSecond(1764547200, None) AND test.ts < TimestampSecond(1767225600, None)\
+        \n    TableScan: test";
+        assert_eq!(expected, plan.to_string());
+
+        // 2025-12-1T00:00:00 <= timestamp < 2025-12-2T00:00:00
+        let sql = "SELECT * FROM test WHERE ts WITHIN '2015-12-1'";
+        let plan = do_query(sql).await.unwrap();
+        let expected = "Projection: *\
+        \n  Filter: test.ts >= TimestampSecond(1448928000, None) AND test.ts < TimestampSecond(1449014400, None)\
+        \n    TableScan: test";
+        assert_eq!(expected, plan.to_string());
+
+        // 2025-12-1T01:00:00 <= timestamp < 2025-12-1T02:00:00
+        let sql = "SELECT * FROM test WHERE ts WITHIN '2025-12-1T01'";
+        let plan = do_query(sql).await.unwrap();
+        let expected = "Projection: *\
+        \n  Filter: test.ts >= TimestampSecond(1764550800, None) AND test.ts < TimestampSecond(1764554400, None)\
+        \n    TableScan: test";
+        assert_eq!(expected, plan.to_string());
+
+        // 2025-12-1T01:12:00 <= timestamp < 2025-12-1T01:13:00
```

Comment:
We can move some cases to the unit test for the `try_to_infer_time_range()`. It will be easier to add more cases.